src/utils/path.c

```diff
@@ -0,0 +1,389 @@
+#include "utils/path.h"
+
+#include "c_string_util.h"
+#include "log.h"
+
+#include <assert.h>
+#include <string.h>
+
+#if defined(_WIN32) || defined(WIN32)
+#include <shlobj.h>
+#include <shlwapi.h>
+#include <windows.h>
+#else
+#include <dirent.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#endif
+
+#define ENSURE_ZERO(str) str[PATH_MAX_LENGTH - 1] = '\0'
+
+static void normalize_slashes(path *p) {
+    char *str = p->buf;
+    while(*str != '\0') {
+        if(*str == '\\') {
+            *str = '/';
+        }
+        str++;
+    }
```

Comment:
Let's not convert `\` to `/` on Linux.
Backslashes are valid within path components there.

As prior work, std::filesystem::path::make_preferred will convert `/` to `\` on Windows, but will not vice-versa on Linux.

Converting `/` to `\` on Windows is largely unnecessary, but it's an aesthetic touch if we need to display a path to the user.

Reply:
I originally added the conversion because there were big phat warnings in stackoverflow that not all windows APIs support the forward slashes. But looking into the docs, it seems all the basic APIs use the same normalization steps, which means the forward slashes should be supported.

I don't know, it might be just easier to just normalize to forward slashes and then use those everywhere :/

Reply:
Decided to drop the windows slash conversion. It's a minor performance optimization to leave it as it is :)